### ISSUE TITLE
Identify builders and properties with solar permits in BZ 

### ISSUE THREAD

**Issue by jcdavis56** [annitodd/PhoenixDataRepo#66]
# Summary 

Goal: We would like to know which builders have experience with solar 

We expect there to be multiple patterns in how rooftop solar permits appear in BZ. We want to be able to identify these rooftop solar permits as well as identifying edge cases where a solar permit may look different from the typical "solar system - roof mount" type of entry. In other words, we want to capture these solar permits that are not easily found by searching for "roof top solar" in the description (or however these permits normally appear). 

# Tasks 

- [x] Create a data dictionary to identify  solar permits
- [ ] Clean solar data to identify rooftop permits 
- [x] Subset data to properties with a builder attatched to it (either in business name or homeowner column)- includes all associated permits with that property
- [ ] Classify solar permits as either belonging to the  builder or likely a retrofit 

# Methodology 

This process is made up of two scripts. A "cleaning" script and a second joining script. 

The first step is to join the columns created in the "PHX_STB_02a_BZ_Name_Matching_All.csv" which are the identified builder permits in BZ from our matching algorithim. I join this dataset to the original raw bz data. Then we want to pull our properties which are associated with a builder. We know that the "PHX_STB_02a_BZ_Name_Matching_All.csv" contains all our builder identified permits, so I store each of the distinct property IDs into a seperate vector. I then use these property IDs to subset the buildzoom data to include all permits which contain the previous vector of builder property ID's. After we subset the data, we should have all the permits which are associated with a property ID which we know to be attatched to a builder. This means that a permit will be included even if the builder is not listed in the homeowner or business_name columns (which might happen with a retrofit for example). 

I then create a data dictionary of words we assume to be commonly associated with solar panels. I collapse all the alike descriptions (from the subset of properties associated with a builder) into a single observation so that we have only the distinct descriptions. I then filter for the descriptions which contain those words commonly related to solar permits. Currently this creates a dataframe of about 13k rows. 

I then try to remove solar permits which we know are not related to rooftop solar such as hot water heaters, pools, and ground installations. The next step identifies permits with a dummy variable which we know for certain relate to rooftop solar. Finally, I seperate the descriptions which must be checked manually in excel into a seperate file. As of right now, this file has roughly 11k rows. Keep in mind that these are the distinct descriptions, so each of these descriptions could be associated with multiple with permits in our bz builder data. A seperate dataset is created in excel which contains two dummy columns for true positive rooftop solar permits and a column for unrelated permits. 

The second script joins the results of the manually identified permits to our bz builder data. The final step in this process is to classify whether a solar permit is either a retrofit or new construction - see issue #65

**Comment by jcdavis56** [comment]
@RobBot00 @wolske @annitodd 

Posting some thoughts on here on this issue since I can't make the normal weekly meetings. 

Was doing some work on building the solar data dictionary to start fitlering for roofotp solar after we have identified the builders. In a previous meeting with @RobBot00 and @wolske, we discussed how filtering for the solar permits would likely be a search for a vector of words we believe will be related to rooftop solar. 

The majority of the bz data is made up of solar permits. We have roughly 4.1 million permits in BZ which are labeled as solar installation. More than likely, there are additional solar permits in other project types such as electrical, new construction, or maybe even roofing. 

The most effecient way to identify these permits might be to search in the description column. This is generally the column which contains the most information about the permit. If we collapse all the alike descriptions to a single observation, we have 1.57 million rows (the number of unique descriptions for all the BZ data). 

If we search for a basic list of words in the description column which contains rooftop, solar, pv, or photovoltaic, then we pick up about 1.4 million rows. Many of these are likely not rooftop solar. Maybe the best approach is to

1) search for the list of words we are confident will be related to a solar permit

then

2) filter out the words we know specifically are not rooftop solar? Might be easier than trying to search out the combination of words we think might contain a rooftop solar permit 

Let me know what you all think!

**Comment by jcdavis56** [comment]
> To clarify, are the 4.1 million solar permits and 1.57 million row numbers based on the data that has been subsetted based on the properties that have builders associated with them? (i.e. property ID's where we have an identified builder)

@RobBot00 

That is just the raw data. Still waiting on the identified builders in the homeowner column I believe

**Comment by eshanprashar** [comment]
## Revised approach

Our goals are: 
1. Identify roof-top solar permits in BuildZoom data
2. Within rooftop solar permits, figure out which ones are for new construction homes 

## High-Level Approach and Key Decisions: 

* **Selecting Columns from BuildZoom data** - BuildZoom data has multiple columns that could be leveraged to achieve goal 1. mentioned above. Some examples are _description_, _type_ and _project_type_. The definitions for these columns are mentioned in this [sheet](https://uchicago.app.box.com/file/1261034262398?s=279onzfabrbg8i20rzp5ywadal1gn36d), which is mentioned in [this Asana task](https://app.asana.com/0/1203068437535585/1205076090350668/f). Looking at BuildZoom data and this data dictionary tells us the following: 
   * the column _type_ has 1K+ categories and these vary according to jurisdiction
   * the column _project_type_ has 31 categories (manageable) but these are defined by BuildZoom according to their 'proprietary' algorithm
   * the column _description_ has 1.5M+ unique values. Values here are supposed to contain detailed information about the 'work' associated with the permit. Our **v1 approach** to find roof-top solar permits should leverage this column as the primary source of information. Our v2, v3 approaches can leverage _project_type_ and _property_ID_ columns.   

* **Selecting Dataset to work with**  - Looking at the description column for:
  * Entire BuildZoom data gives us 1.5M+ unique values 
  * Our matched builder dataset with ~100K rows gives us ~16K unique values
  
  **[Key decision]:** 
  * If we want to look into other non-builder matched rows to get roof-top permits, we should work with the entire BuildZoom data set with 1.5M+ values and leverage Machine Learning to classify _rooftop_solar_ as 0 / 1. The benefit here will be our approach will be more generic, the major con will be that it may require multiple iterations.  
  
  * If we are confident that solar permit data will not be outside the confines of the matched ~100K rows dataset, then we should work with ~16K descriptions. In this case, Machine Learning is not really needed - we probably just need to filter for roof-top solar and in cases where that doesn't work, apply some logic on **number of permits for a given property ID** (for example if a property has a permit for _solar_installation_ and _new_construction_, then we may mark it 1 even with a vague description like **Solar**) 

* **Building ML model** - For now, I have explored the idea of using _ 1.5M+ descriptions_ to classify _rooftop_solar_ as 0 / 1. I used this starting point because of 3 reasons:  
   * As mentioned above, this approach is more generic
   * An ML model would need us to manually classify **some** descriptions using which, it will classify the others for us. Looking at the frequency distribution of descriptions, ~600 account for 1.3M BuildZoom rows (17% of the dataset). If we can label these 600 manually, we should have a good starting point for patterns for _rooftop_solar = 0 or 1_. 
   * We can build upon an ML model by adding more features based on other logic. For example, we can add a feature called _has_builder_ which can be 0 / 1 depending on whether a description has one of the ~470 builder names in it. 

* **Other ways to assign rooftop_solar** : After using descriptions, we can turn towards the _project_type_ column. For a given property ID, if the permit description has something vague like _solar_ or _pv system_, we can take a look at what values exist in the _project_type_ for that property_ID. Possible values include: solar installation, new construction, electrical work, roofing, commercial renovation etc. 

@RobBot00 : For now, I have 2 questions: 
1. Is it fair to assume that we might be interested in finding solar permit in the dataset outside of 100K rows or is that not the case?
2. Do you have any ideas around using _project_type_ column to assign _rooftop_solar_ 0 / 1? 

Please let me know in case of any doubts or questions!

**Comment by RobBot00** [comment]
AWESOME summary Eshan. @wolske and @annitodd, I suggest reading Eshan's comment for a refresher on what we are doing with the BZ data currently; also please continue to read this comment.

@eshanprashar quick clarification: the 100k rows from our matched builder set, is that filtering based on which observations have a builder in Business_name or Homeowner_name? OR is it returning all observations with property ID's that have a Business_name or Homeowner_name? It should be the latter, I think it is but want to double check! 

To your questions:

> Is it fair to assume that we might be interested in finding solar permit in the dataset outside of 100K rows or is that not the case?

Nope! We are only interested in classifying the 100k.

> Do you have any ideas around using project_type column to assign rooftop_solar 0 / 1?

Ehh... From my experience, I found project_type to represent permits for things like solar water heating, solar storage systems, and a few other miscellaneous items. I think we should start with descriptions and see where we get. I like your V1, V2, and V3 approach.  It really just depends on how conservative we want to be, let me explain:

**Careful approach**: Use the description as a whitelist; if keywords like "Rooftop solar", "PV System", etc. show up in the description, then assign rooftop_solar to 1. This approach would involve building a dictionary of words we feel confident are rooftop solar.

**Loose Cannon approach**: We could instead use the description as a blacklist. Here, we would filter down to project_type = solar, then create a dictionary of keywords that aren't rooftop solar like "water heater", "ground mount", etc. The idea here is that we trust BZ's categorization of permit type, and so ones that don't include our keywords we are assuming are rooftop solar.

@annitodd and @wolske, please let me know if you have thoughts on this.

Regardless, I think a bag-of-words exercise is a good first step regardless of the approach.

**Comment by eshanprashar** [comment]
> @eshanprashar quick clarification: the 100k rows from our matched builder set, is that filtering based on which observations have a builder in Business_name or Homeowner_name? OR is it returning all observations with property ID's that have a Business_name or Homeowner_name? It should be the latter, I think it is but want to double check!

100K rows are in the the former - if I change it to the latter, there should be potentially more rows. I'll get back on the exact number. 


> To your questions:
> 
> > Is it fair to assume that we might be interested in finding solar permit in the dataset outside of 100K rows or is that not the case?
> 
> Nope! We are only interested in classifying the 100k.

Okay cool! Then I'll just work with the descriptions associated with property IDs of the matched rows.

**Comment by RobBot00** [comment]
> These could be ground-mounted solar systems. How should we think about these? Is there some other column in BuildZoom that can guide in such cases?

Emphasis on _could_ haha. Based on LBNL's research, is is exceedingly rare to see ground mount on new construction. They find ~160 across 15+ states. In short, if we know its PV panels, we can feel fairly confident that it is roof mounted. That said, if some specify ground mounting, we can definitely classify those as not rooftop solar.

**Comment by eshanprashar** [comment]
> > These could be ground-mounted solar systems. How should we think about these? Is there some other column in BuildZoom that can guide in such cases?
> 
> Emphasis on _could_ haha. Based on LBNL's research, is is exceedingly rare to see ground mount on new construction. They find ~160 across 15+ states. In short, if we know its PV panels, we can feel fairly confident that it is roof mounted. That said, if some specify ground mounting, we can definitely classify those as not rooftop solar.

Okay great!

**Comment by wolske** [comment]
With regard to this conversation:
> > @eshanprashar quick clarification: the 100k rows from our matched builder set, is that filtering based on which observations have a builder in Business_name or Homeowner_name? OR is it returning all observations with property ID's that have a Business_name or Homeowner_name? It should be the latter, I think it is but want to double check!
> 
> 100K rows are in the the former - if I change it to the latter, there should be potentially more rows. I'll get back on the exact number.

Just want to affirm what Robert said. We want to subset on **all** permits associated with a _property_  that has been matched to a production builder -- not just the permits that have builder in the business or homeowner name. It is possible, for example, that a production company could sell a home while it is still being built. In that case, by the time solar is added, the permit might have the solar installer as business name and the real homeowners as homeowner. Once we identify that the permit tagged as solar is actually solar PV, our second task (as you already defined it above) would be to scrutinize whether this solar permit is a retrofit or part of a new build. 

@eshanprashar can you share what words you've been using in the dictionary? kW and killowatt should also be included.

**Comment by annitodd** [comment]
**suggestion of v1a approach**
@eshanprashar 
Is there a list of the 31 categories in project_type (I didn't see it in the sheet):
> * the column _project_type_ has 31 categories (manageable) but these are defined by BuildZoom according to their 'proprietary' algorithm

A different approach -- v1a, say -- would be relatively quick, may result in an obvious easy classification (or might be messy in which case we could revert to v2): run a classification of the variables _description_ and _type_ on _project_type_ -- that would point out if like, how good or bad BZ classification is.

**Different approach v10a**
@eshanprashar do you know how many of our production builders have had at least one match in the business or homeowner name?

Asking because a totally different approach --would be to 
1. for each of our production builders, count how many properties in the BZ dataset are associated with that builder
2. Identify the production builders that have only a few properties associated with them (like, are there any prod builders that have less than 5 properties associated? What is the smallest amount?)
3. For those prod builders with only a few properties, examine the three vars (type, descr, proj type), see if any of the permits are obviously solar (maybe cluster the three vars)

**Comment by annitodd** [comment]
@eshanprashar 
Actually a rough guess of the number of our builders that have matched at least once in the BZ dataset would also be super helpful for the survey development.  (even if the match isn't perfect). If there's any chance  you could do a rough estimate of that it'd be great!

**Comment by eshanprashar** [comment]
> @eshanprashar - as always, great job writing this up in an easy-to-follow way!
> 
> **On the solar dictionary/set:**
> 
> * I think the list make sense based on your analysis. Maybe add "panel" (singular form)
> * You mentioned that "killowatt" was not in the dictionary. There's a typo there. Can you check if kilowatt shows up anywhere? I wouldn't be surprised to see it as kilo watt, too, even though that's not technically right.
> * I noticed that "roof" by itself is in the dictionary. Is it possible we're picking up reroofing/new roof projects that don't involve solar?
> * It's interesting that so many combinations related to new single family homes come up in the n-gram analysis. We should remember this when determining which solar permits are actually new construction solar.

Thanks for those inputs @wolske! I'll remove 'roof' and see how the results change. I'll add 'panel' and 'kilowatt' (which does exist in descriptions) and/or different versions of it. 


> **On unmatched solar rows:**
> 
> * My general hunch is that at least for some of these permits, the solar permit is pulled as part of a larger project (which might explain why the project type doesn't seem to match). As an example, I had a kitchen remodel on a prior home where we ended up putting a new roof on part of our house. I wouldn't be surprised if the project type for the roof was listed as kitchen remodel as that's what the general contractor was hired to do.
> * Checking for spelling errors on solar terms is a good start.  I have no idea what to do with descriptions like the example you gave (bp 882306 tmk 91017206 911365 kakiwi st lot 28 4 787 sf hoopili ph13 parcel 98 hooulu ph 2 plan 2r new single family dwelling w air conditioning unit).
> * for the records that BZ says are solar but have blank descriptions, can we check the business name to see if they appear to be solar installers?
> * Can you do another n-gram analysis of the unmatched rows so that we can get a sense of the descriptors that are coming up most frequently? That might help us understand what's going on.
> 
> There are probably a variety of patterns that we could try to identify by eyeballing/manually checking the data. But I'd like to go back to your idea of using Machine Learning across the entire dataset. Do you have a sense of how long that would take? And what's your hunch on how the end result would compare to what you've already done?

I'll add the n-grams for non-matched rows and check the business names (by end of today/ first half tomorrow). 

Finally, your question around efficiency and effectiveness of using Machine Learning is a tough one to be honest. The short answer is there are 2-3 algorithms we can try, and each can take us ~3-4 days to tune because of the following 2 aspects: 

1. **Input data / features for training**: All Machine Learning algorithms will take one thing as an input, manually-labelled data. We can use the 0/1 we've assigned using our dictionary/ set as a starting point. However, 100K labelled rows are extremely few when we're dealing with a dataset that has ~7M rows (entire BZ dataset) and want to make predictions for that many rows. It is ideal to label 40-50% of the data to have predictions that make sense. To make up for this deficit, we can:
   * figure out a way to label more rows 
   * in addition to description, can potentially add **more features** in the algorithm (for example _Property Type_ or a combination of some columns) or
   * eliminate rows based on some criteria so that our dataset size reduces

2. **Balancing classes in outcome variable**: In the manually labelled data, we should ideally have equal or 65-35 representation from both classes - [0 = not a permit and 1 = permit]. This is neither true in our dataset of ~120K rows nor will likely be true for the entire dataset of 7M rows. There are workarounds for this problem, but I'll have to read up a bit on that topic. 

As we iterate over these 2 things above, we will monitor performance-metrics associated with these ML algorithms. The metrics are pretty standard and will give us a clear answer on model performance. Actually coding the model and monitoring / interpreting the metrics is the easy part - steps 1 /2 will likely take the most time because typically, there is no straight-forward answer unless there is published work on a similar dataset.

**Comment by eshanprashar** [comment]
> @eshanprashar - as always, great job writing this up in an easy-to-follow way!
> **On the solar dictionary/set:** 
> * I think the list make sense based on your analysis. Maybe add "panel" (singular form)
> * You mentioned that "killowatt" was not in the dictionary. There's a typo there. Can you check if kilowatt shows up anywhere? I wouldn't be surprised to see it as kilo watt, too, even though that's not technically right.
> * I noticed that "roof" by itself is in the dictionary. Is it possible we're picking up reroofing/new roof projects that don't involve solar?
> * It's interesting that so many combinations related to new single family homes come up in the n-gram analysis. We should remember this when determining which solar permits are actually new construction solar.
> 
> Thanks for those inputs @wolske! I'll remove 'roof' and see how the results change. I'll add 'panel' and 'kilowatt' (which does exist in descriptions) and/or different versions of it.

* Upon adding _panel_ and _kilowatt_, out of 117,235, 94,343 rows have a '1' for rooftop_solar. Earlier, this number was  94,240, so an addition of 103 rows.
* After removing the token _roof_, rows with _rooftop_solar = 1_ reduce to 93,860 

> I'll add the n-grams for non-matched rows and check the business names (by end of today/ first half tomorrow).

Here are the n-grams for rooftop_solar = 0 rows (23,393 rows): 
* **Uni-grams, top 15** - ('tmk',), 12321), (('bp',), 11557), (('lot',), 10701), (('family',), 10453), (('single',), 10307), (('unit',), 10130), (('new',), 9584), (('plan',), 9552), (('w',), 9235), (('dwelling',), 8108), (('parcel',), 7794), (('air',), 7297), (('conditioning',), 7288),(('4',), 6145), (('st',), 5634)...

* **Bi-grams, top 10** - [(('single', 'family'), 10303), (('new', 'single'), 8823), (('family', 'dwelling'), 7545), (('air', 'conditioning'), 7288), (('conditioning', 'unit'), 7143), (('w', 'air'), 6660), (('dwelling', 'w'), 6549), (('st', 'lot'), 3550),(('ph10a', 'parcel'), 3210), (('lp', 'lot'), 2775)...

* **Tri-grams, top 10** - [(('new', 'single', 'family'), 8823), (('single', 'family', 'dwelling'), 7515), (('air', 'conditioning', 'unit'), 7143), (('w', 'air', 'conditioning'), 6660), (('family', 'dwelling', 'w'), 6102), (('dwelling', 'w', 'air'), 6081), (('ph', '2', 'plan'), 1830), (('honouliuli', 'st', 'unit'), 1695), (('conditioning', 'unit', 'former'), 1602), (('sfhoopili', 'ph10a', 'parcel'), 1560),

* **Quad-grams, top 10** - [(('new', 'single', 'family', 'dwelling'), 7280), (('w', 'air', 'conditioning', 'unit'), 6519),
 (('dwelling', 'w', 'air', 'conditioning'), 6081), (('single', 'family', 'dwelling', 'w'), 6072), (('family', 'dwelling', 'w', 'air'), 5940), (('air', 'conditioning', 'unit', 'former'), 1602), (('ph1', 'parcel', '1', '4'), 1386), (('parcel', '1', '4', 'plan'), 1386),(('hoopili', 'ph1', 'parcel', '1'), 1271), (('sf', 'hoopili', 'ph1', 'parcel'), 1181)...

**Comment by RobBot00** [comment]
@eshanprashar To this concern: 

>  However, 100K labelled rows are extremely few when we're dealing with a dataset that has ~7M rows (entire BZ dataset)

Just wanted to chime in: at this stage, we are only interested in identifying solar permits for properties that have builders associated with them; so this should be significantly less (however still a lot). The other aspect here is that there should be significantly less unique descriptions. I'm curious, how many **unique** permit descriptions do we have for the subset of observations that have builders associated with properties?

One idea here is that we use the "solar" and "photovoltaic" exact matches as the true positives; however relying too heavily on these kinds of techniques I imagine would overfit the model on your key words. Thoughts?

**Comment by eshanprashar** [comment]
> @eshanprashar To this concern:
> 
> > However, 100K labelled rows are extremely few when we're dealing with a dataset that has ~7M rows (entire BZ dataset)
> 
> Just wanted to chime in: at this stage, we are only interested in identifying solar permits for properties that have builders associated with them; so this should be significantly less (however still a lot). The other aspect here is that there should be significantly less unique descriptions. I'm curious, how many **unique** permit descriptions do we have for the subset of observations that have builders associated with properties?

@RobBot00 : 
* The dataset with builders associated with them has ~18K unique descriptions
* The BuildZoom dataset overall has ~1.1M descriptions. 

It is unlikely that using Machine Learning for ~18K descriptions will be a good investment of time because the additional time investment will not give additional insights on data. From the little that I know about ML, these learning algorithms usually helps with larger datasets (for example for 1.1M descriptions).    

> One idea here is that we use the "solar" and "photovoltaic" exact matches as the true positives; however relying too heavily on these kinds of techniques I imagine would overfit the model on your key words. Thoughts?

I'm not sure I understand the meaning of using _'solar' exact match as true positive_. We can discuss this during our call today.

**Comment by RobBot00** [comment]
> We can discuss this during our call today.

@eshanprashar OH MAN! Did we miss a call with you yesterday?! I don't see anything in my calendar, please let us know!

@wolske
